### ISSUE TITLE
[question] Load balancing with KCL enabled

### ISSUE THREAD

**Issue by v9ru** [spring-cloud/spring-cloud-stream-binder-aws-kinesis#121]
At the end of Kinesis Binder overview section in the kinesis binder [documentation](https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/blob/master/spring-cloud-stream-binder-kinesis-docs/src/main/asciidoc/overview.adoc#kinesis-binder-overview) I see the following

> On the consumer side the instanceCount and instanceIndex are used to distribute shards between consumers in group evenly.

In KCL [documentation](https://docs.aws.amazon.com/streams/latest/dev/developing-consumers-with-kcl.html) I see the following

> You can run a KCL application on any number of instances. Multiple instances of the same application coordinate on failures and load balance dynamically

So does this mean, If I enable KCL with binder I can ignore the instanceCount and instanceIndex properties and get load balancing?

**Comment by artembilan** [comment]
That's correct. 
Let me fix that doc! It is really relevant to our own balancing.
In case of KCL the environment is fully out of our control.
Plus bear in mind that KCL cal deal only with a single stream and there is fully no any way to assign specific shards to the consumer.

In case of plain Kinesis consumer implementation we have a fully control with those `instanceCount` and `instanceIndex` to select and asign specific shards to the consumer during configuration phase.

**Comment by v9ru** [comment]
@artembilan It is not working as expected.
I have created a stream with 4 shards and started 2 consumers and I have the following configuration

```
spring:
  profiles: env-local

  cloud:
    stream:
      bindings:
        output:
          destination: nvivek-kinesis-binder-delete-this
          content-type: text/plain

        input:
          destination: nvivek-kinesis-binder-delete-this
          group: nvivek-kinesis-binder-delete-this-group
          content-type: text/plain

      kinesis:
        binder:
          autoCreateStream: false
          kplKclEnabled: true

          checkpoint:
            table: nvivek-kinesis-binder-delete-this-metadata-store

          locks:
            table: nvivek-kinesis-binder-delete-this-lock-store

        bindings:
          input:
            consumer:
              startTimeout: 6000
              listenerMode: record
              checkpointMode: record
              workerId: nvivek
              idleBetweenPolls: 1000
              consumerBackoff: 1000
              shardIteratorType: TRIM_HORIZON
```
When I start 1st consumer, it gets all the 4 shards and when I start 2nd consumer it takes control on all the 4 shards.

I see the following logs in the 1st consumer

```
2020-01-10 14:13:52.771 [ INFO] 23883 --- [aseRenewer-0003] ervices.kinesis.leases.impl.LeaseRenewer : Worker nvivek lost lease with key shardId-000000000000
2020-01-10 14:13:52.773 [ INFO] 23883 --- [aseRenewer-0004] ervices.kinesis.leases.impl.LeaseRenewer : Worker nvivek lost lease with key shardId-000000000003
2020-01-10 14:13:52.775 [ INFO] 23883 --- [aseRenewer-0000] ervices.kinesis.leases.impl.LeaseRenewer : Worker nvivek lost lease with key shardId-000000000001
2020-01-10 14:13:52.775 [ INFO] 23883 --- [aseRenewer-0002] ervices.kinesis.leases.impl.LeaseRenewer : Worker nvivek lost lease with key shardId-000000000002
2020-01-10 14:13:53.309 [ INFO] 23883 --- [askExecutor-222] d.kinesis.KclMessageDrivenChannelAdapter : Scheduler is shutting down for reason 'ZOMBIE'; checkpointing...
2020-01-10 14:13:53.309 [ INFO] 23883 --- [askExecutor-223] d.kinesis.KclMessageDrivenChannelAdapter : Scheduler is shutting down for reason 'ZOMBIE'; checkpointing...
2020-01-10 14:13:54.311 [ INFO] 23883 --- [askExecutor-224] d.kinesis.KclMessageDrivenChannelAdapter : Scheduler is shutting down for reason 'ZOMBIE'; checkpointing...
2020-01-10 14:13:54.311 [ INFO] 23883 --- [askExecutor-225] d.kinesis.KclMessageDrivenChannelAdapter : Scheduler is shutting down for reason 'ZOMBIE'; checkpointing...
```
Am I missing any configuration?

**Comment by artembilan** [comment]
Well, one more time: when we use KCL, we only provide an `applicationId` and a `streamName`. All the hard work about balancing is done outside of the binder.
Unfortunately I don't have enough knowledge about KCL to help you to understand what is going on and why your shards are not distributed between instances.
If you wouldn't use KCL, I could say that your concern is very similar to what we have with regular Kinesis consumer implementation...

Only what I can tell about your configuration that you don't need `checkpoint` and `locks` section when you use `kplKclEnabled = true`.

May be `workerId` has to be unique for every instance?..

**Comment by artembilan** [comment]
Forgot to say: try to seek the help about KCL rebalancing exactly in the KCL community. 
This project has nothing how to control KCL behavior.

**Comment by v9ru** [comment]
Thanks @artembilan

I have gone through the KCL docs and workerId has to be unique per consumer (so it will be same for every instance of a consumer).

I have removed checkpoint and locks configuration but it still creates default dynamodbs for checkpoint and locks .. can you make an improvement to avoid that if those tables are of no use when kcl is enabled.

**Comment by artembilan** [comment]
Well, there is an `applicationId` which indicates a consumer group abstraction, but I'm not sure about `workerId`. Its purpose might be really a unique instance per cluster and if we start a new instance with the same `workerId`, an old one dies.
But this is still my guess. Otherwise I don't see a difference between `applicationId` and  `workerId`.
Sorry for my ignorance: I'm not a genuine source of KCL/KPL truth.
Please, consult with them. When get an answer, I'd be glad to learn it for myself, too. But that's different story...

> can you make an improvement to avoid that if those tables are of no use when kcl is enabled.

I think I did it recently: https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/issues/117

Please, consider to upgrade to just released Kinesis Binder `2.0.1`: https://spring.io/blog/2020/01/10/spring-integration-aws-2-3-1-spring-cloud-stream-kinesis-binder-2-0-1-available

**Comment by v9ru** [comment]
Sorry, my bad. I understood it wrong. 
You are correct, applicationId is used for consumer group abstraction.

I removed the workerId property from my configuration and now I see the load balancing.
It explains my initial problem too where a single instance is taking control of all shards.

Thanks a lot for all the help @artembilan
